Mathlib/NumberTheory/FLT/Three.lean

```diff
@@ -342,6 +364,292 @@ lemma exists_Solution_of_Solution' : ∃ (S₁ : Solution hζ), S₁.multiplicit
+    ⟨2 * S.b, by ring⟩
+
+/-- Given `(S : Solution)`, we have that `λ ^ 2` does not divide `S.a + η * S.b`. -/
+lemma lambda_sq_not_a_add_eta_mul_b : ¬ λ ^ 2 ∣ (S.a + η * S.b) := by
+  simp_rw [a_add_eta_mul_b, dvd_add_right S.hab, pow_two, mul_dvd_mul_iff_left
+    hζ.zeta_sub_one_prime'.ne_zero, S.hb, not_false_eq_true]
+
+/-- Given `(S : Solution)`, we have that `λ ^ 2` does not divide `S.a + η ^ 2 * S.b`. -/
+lemma lambda_sq_not_dvd_a_add_eta_sq_mul_b : ¬ λ ^ 2 ∣ (S.a + η ^ 2 * S.b) := by
+  intro ⟨k, hk⟩
+  rcases S.hab with ⟨k', hk'⟩
+  refine S.hb ⟨(k - k') * (-η), ?_⟩
+  rw [show S.a + η ^ 2 * S.b = S.a + S.b - S.b + η ^ 2 * S.b by ring, hk',
+    show λ ^ 2 * k' - S.b + η ^ 2 * S.b = λ * (S.b * (η +1) + λ * k') by rw [coe_eta]; ring,
+    pow_two, mul_assoc] at hk
+  simp only [mul_eq_mul_left_iff, hζ.zeta_sub_one_prime'.ne_zero, or_false] at hk
+  apply_fun (· * -↑η) at hk
+  simp only at hk
+  have := hζ.toInteger_isPrimitiveRoot.isRoot_cyclotomic (by decide)
+  simp only [PNat.val_ofNat, Polynomial.cyclotomic_three, Polynomial.IsRoot.def,
+    Polynomial.eval_add, Polynomial.eval_pow, Polynomial.eval_X, Polynomial.eval_one] at this
+  rw [show (S.b * (η + 1) + λ * k') * -η = (- S.b) * (η ^ 2 + η + 1 - 1) - η * λ * k' by ring,
+    coe_eta, this, zero_sub, mul_neg, mul_one, neg_neg, sub_eq_iff_eq_add] at hk
+  rw [hk, coe_eta]
+  ring
+
+lemma eta_add_one_mul_neg_eta_eq_one : ((η : 𝓞 K) + 1) * (-η) = 1 :=
+  calc ((η : 𝓞 K) + 1) * -η = -(η ^ 2 + η + 1) + 1 := by ring
+  _ = 1 := by
+    ext
+    have := hζ.isRoot_cyclotomic (by decide)
+    simp only [PNat.val_ofNat, Polynomial.cyclotomic_three, Polynomial.IsRoot.def,
+      Polynomial.eval_add, Polynomial.eval_pow, Polynomial.eval_X, Polynomial.eval_one] at this
+    simp only [IsUnit.unit_spec, RingOfIntegers.neg_mk, neg_add_cancel_comm, map_add,
+      RingOfIntegers.map_mk, map_neg, map_pow, map_one, this, neg_zero, zero_add]
+
+/-- If `p : 𝓞 K` is a prime that divides both `S.a + S.b` and `S.a + η * S.b`, then `p`
+is associated with `λ`. -/
+lemma associated_of_dvd_a_add_b_of_dvd_a_add_eta_mul_b {p : 𝓞 K} (hp : Prime p)
+    (hpab : p ∣ S.a + S.b) (hpaetab : p ∣ S.a + η * S.b) : Associated p λ := by
+  suffices p_lam : p ∣ λ from hp.associated_of_dvd hζ.zeta_sub_one_prime' p_lam
+  by_contra p_lam
+  refine hp.not_unit <| IsCoprime.isUnit_of_dvd' S.coprime ?_ ?_
+  · refine (hp.dvd_or_dvd ?_).resolve_left ‹_›
+    rw [show λ * S.a = η * (S.a + S.b) - (S.a + η * S.b) by rw [coe_eta]; ring]
+    exact dvd_sub (dvd_mul_of_dvd_right hpab _) hpaetab
+  · refine (hp.dvd_or_dvd ?_).resolve_left ‹_›
+    rw [show λ * S.b = (S.a + η * S.b) - (S.a + S.b) by rw [coe_eta]; ring]
+    exact dvd_sub hpaetab hpab
+
+/-- If `p : 𝓞 K` is a prime that divides both `S.a + S.b` and `S.a + η ^ 2 * S.b`, then `p`
+is associated with `λ`. -/
+lemma associated_of_dvd_a_add_b_of_dvd_a_add_eta_sq_mul_b {p : 𝓞 K} (hp : Prime p)
```

Comment:
What about a general lemma that says something like "`p ∣ a*x + b*y` and `p ∣ c*x + d*y` imply that `p` divides `(a*d-b*c) * gcd x y`" ? That could simplify this proof and the next.